Implementing when guards as discussed

FraMecca opened the pull request:
Guards still need to be handled at the parser level.
In particular I think we need to catch the entire (if (apply ... s-expression

Also, this PR made me reflect on the fact that source_blackbox and target_blackbox, even if they are both strings, are not going to work with string equality but need something more sofisticated

Commit:
trivial guards example

Commit:
Add AST node for (if (apply ... lambda code to handle guards

Commit:
ignore guards in the target to target equivalence checking algorithm

Commit:
Guard node in the constraint trees

Source_sym_engine and Target_sym_engine and Merge_accessors

Commit:
handle guards in equivalence checking by a queue of blackbox

gasche commented:
You are right, string equality is not the right way to check for equivalence of blackboxes.

I would propose to use another hack, similar to what I suggested for guards: let's restrict the structure of blackboxes we are considering in our prototype, so that we can extract more structure from them and can check equivalence of source and target blackboxes by inspecting the structure.

We could make all our input code look like the following:

```ocaml
external guard : 'a -> 'b = "guard"
external observe : 'a -> 'b = "observe"

[@@@warning "-20"] (* Warning 20: Unused function argument *)
(* this attribute silences a warning you get when you try to apply any
   of the primitives above to several arguments (this is type-correct
   but we know from the polymorphism that the arguments are ignored) *)

type t = Z | S of t

let _ = function
  | Z -> observe 0
  | S Z -> observe 1
  | S x when guard x -> observe 2
  | S (S x) as y when guard x y -> observe 3
  | S _ -> observe 4
```

Namely:
- we assume two external functions `guard` and `observe` with a weird type that lets us pass any number of arguments to them
- all our guards are of the form `guard <arg> <arg> <arg>`, where the `<arg>` are in the subset of the grammar that we know how to symbolically evaluate (for example: formed of value constructors and pattern variables)
- similarly, all our right-hand-sides are of the form `observe <arg> <arg> ...` with the same constraints on arguments

With this restriction it is easy to recognize and "parse" guards or leaves (get the list of symbolic values of arguments), at both the source and the target level, and we can check that they are equivalent.

gasche commented:
I looked at the PR and I think the code is fine -- it's not complete as there is no parsing support but the parts that are there look good. Github may be lying about the commit order, but it looks you commited the `| Guard .. -> failwith "not implemented"` changes *before* you defined the `Guard` constructor, so the intermediate step does not build, which is a bit weird). I was going to say that I have nothing to nitpick about, but I changed my mind: I would rename `IfApplỳ` into `IfGuard` (or just `Guard` but maybe you want to avoid the name conflict).

Commit:
renamed ifapply to ifguard as of comments to pr

Commit:
adopted proposal for handling blackboxes

Commit:
Target blackbox refactored

Commit:
to rebase

gasche committed:
parser support for 'observe' in target programs

gasche committed:
correctly handle numeric overflows when shifting domains

gasche committed:
finish parser support and convert all examples

gasche committed:
support for 'observe' in the source parser

gasche committed:
parse guards in the source examples as well

gasche committed:
propagate guards during matrix decomposition

gasche committed:
generate guards in source constraint trees

gasche committed:
fix the printing of guards in constraint trees

Commit:
Merge pull request #24 from gasche/guard-observe-parser

Target parser support for `guard` and `observe`

Commit:
Merge pull request #25 from gasche/source-guard-sym-exec

Guards in source program and symbolic execution

Commit:
Merge branch 'master' of github.com:FraMecca/inria-internship into when-guards

Commit:
Merge branch 'when-guards' of github.com:FraMecca/inria-internship into when-guards

gasche committed:
fix a small equivalence-checking bug

The test in main.ml:

    assert (Equivalence.compare result result)

would fail on most programs because of a subtle bug in the equivalence
checker: it would only kill branches whose domain is empty *before*
trimming, instead of testing for emptiness after computing the
intersection with the trimming condition.

Commit:
implement canonical_form for comparison between sym_values
Added Sym_values module

gasche committed:
fix a but in target_sym_exec

The following program would fail with a Not_found exception

  let
    ((a = 1)
     (b = a + 2))
  in
    (a + b)

because both bindings would be evaluated in the initial environment,
instead of having 'a' defined when evaluating 'b'.

gasche commented on engine/sym_values.ml:
```diff
@@ -0,0 +1,74 @@
+      | None -> assert false
+  in
+  let rec compare_canonical_form_ : source_canonical_form * target_canonical_form -> bool = function
+    | (NonSingleton s, NonSingleton t) -> s = t (*TODO: DISCUSS: is it correct to ignore pi (domain) *)
```
We should use `Domain.equal` here instead of polymorphic equality.

gasche commented on engine/sym_values.ml:
```diff
@@ -0,0 +1,74 @@
+  in
+  compare_canonical_form_ ((canonical_form_of_source_sym_value src), (canonical_form_of_target_sym_value tgt))
+
+let compare_sym_values find_constructor_of find_accessor_of sym_values target_values =
```
Instead of parametrizing on functions I would parametrize over the corresponding data: the `repr_env` and a `sym_env`.

Commit:
removed Tag from canonical_form
